connectors/sources/jira.py

```diff
@@ -0,0 +1,467 @@
+# or more contributor license agreements. Licensed under the Elastic License 2.0;
+# you may not use this file except in compliance with the Elastic License 2.0.
+#
+"""Jira source module responsible to fetch documents from Jira on-prem or cloud server.
+"""
+import asyncio
+import os
+from copy import copy
+from functools import partial
+from urllib import parse
+
+import aiofiles
+import aiohttp
+from aiofiles.os import remove
+from aiofiles.tempfile import NamedTemporaryFile
+
+from connectors.logger import logger
+from connectors.source import BaseDataSource
+from connectors.utils import (
+    TIKA_SUPPORTED_FILETYPES,
+    ConcurrentTasks,
+    MemQueue,
+    convert_to_b64,
+    iso_utc,
+)
+
+IS_CLOUD = True
+RETRY_INTERVAL = 2
+DEFAULT_CONTENT_EXTRACTION = True
+DEFAULT_SSL_DISABLED = True
+RETRIES = 3
+FILE_SIZE_LIMIT = 10485760
+
+FETCH_SIZE = 100
+CHUNK_SIZE = 1024
+QUEUE_SIZE = 1024
+QUEUE_MEM_SIZE = 5
+MAX_CONCURRENCY = 5
+MAX_CONCURRENT_DOWNLOADS = 50  # Max concurrent download supported by jira
+
+PING = "ping"
+PROJECT = "project"
+ISSUES = "all_issues"
+ISSUE_DATA = "issue_data"
+ATTACHMENT_CLOUD = "attachment_cloud"
+ATTACHMENT_SERVER = "attachment_server"
+URLS = {
+    PING: "/rest/api/2/myself",
+    PROJECT: "/rest/api/2/project?expand=description,lead,url",
+    ISSUES: "/rest/api/2/search?maxResults={maxResults}&startAt={startAt}",
+    ISSUE_DATA: "/rest/api/2/issue/{id}",
+    ATTACHMENT_CLOUD: "/rest/api/2/attachment/content/{attachment_id}",
+    ATTACHMENT_SERVER: "/secure/attachment/{attachment_id}/{attachment_name}",
+}
+
+
+class JiraDataSource(BaseDataSource):
+    """Jira"""
+
+    name = "Jira"
+    service_type = "jira"
+
+    def __init__(self, configuration):
+        """Setup the connection to the Jira
+
+        Args:
+            configuration (DataSourceConfiguration): Object of DataSourceConfiguration class.
+        """
+        super().__init__(configuration=configuration)
+        self.is_cloud = self.configuration["is_cloud"]
+        self.host_url = self.configuration["host_url"]
+        self.ssl_disabled = self.configuration["ssl_disabled"]
+        self.certificate = self.configuration["ssl_ca"]
+        self.enable_content_extraction = self.configuration["enable_content_extraction"]
+        self.retry_count = self.configuration["retry_count"]
+        self.concurrent_downloads = self.configuration["concurrent_downloads"]
+
+        self.ssl_ctx = False
+        self.session = None
+        self.queue = MemQueue(
+            maxsize=QUEUE_SIZE, maxmemsize=QUEUE_MEM_SIZE * 1024 * 1024
+        )
+        self.fetchers = ConcurrentTasks(max_concurrency=MAX_CONCURRENCY)
+
+    @classmethod
+    def get_default_configuration(cls):
+        """Get the default configuration for Jira
+
+        Returns:
+            dictionary: Default configuration.
+        """
+        return {
+            "is_cloud": {
+                "value": IS_CLOUD,
```

Comment:
This default value should be inlined

Reply:
We are following a best practice to create a constant for the default values since that's not gonna change and pass it to configurations. So, do you want us to eliminate the constant and directly pass the default value i.e. True for `is_cloud`?

Reply:
My thought process here:

1. These default-value constants are used in a single place - in `get_default_configuration` method, so they are not re-usable
2. If your IDE does not support tooltips (I use vim, so I'm biased here), then to figure out the value of constant while looking at the configuration, you'll need to scroll to the beginning of file to see the value - that breaks the flow of thoughts/reading slightly.

All in all I just don't see a benefit of moving it to the beginning of the file in a constant instead of using default value. Think of it from this perspective - why labels of configurable fields are not constants too? Why field names are not constants?

To me out of these 3 only field names are worth being constants (as you actually use them in code in multiple places), and all others can just be regular values instead of constants.

Reply:
Cool, Got it. I'll make the configurations inline.